### PR TITLE
[POC] Add API tests package

### PR DESCRIPTION
**Pull request by yellowee**

### CODE REVIEW

File: api-tests/vitest.config.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { defineConfig } from 'vitest/config'
+
+export default defineConfig({
+  test: {
+    include: [__dirname + '*/tests/**/*.test.ts'],
```
Comment (krzysztofzuraw) [inline api-tests/vitest.config.ts:5]:
```suggestion
	include: ["src/**/*.test.{ts,tsx}"],
	globals: true,
```

File: api-tests/tests/test.test.ts
Comment (krzysztofzuraw) [inline api-tests/tests/test.test.ts:1]: **suggestion**: what about keeping those tests under `api-tests/src`?

File: api-tests/tests/test.test.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import { describe, expect, it } from 'vitest'
+
+describe('1st test', () => {
```
Comment (krzysztofzuraw) [inline api-tests/tests/test.test.ts:3]: **question**: shouldn't this test cover a more realistic test scenario?

File: .github/workflows/apiTests.yml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+on:
+  pull_request:
+    branches: [ "main" ]
+
+jobs:
+  tests:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-node@v3
```
Comment (krzysztofzuraw) [inline .github/workflows/apiTests.yml:12]: **suggestion**: you can add here support for `.nvmrc` file

https://github.com/actions/setup-node/blob/main/docs/advanced-usage.md#node-version-file

File: .github/workflows/apiTests.yml [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+      - run: npm i
```
Comment (krzysztofzuraw) [inline .github/workflows/apiTests.yml:13]:
```suggestion
      - run: npm ci
```

File: api-tests/package.json [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  "main": "index.js",
+  "scripts": {
+    "api-tests": "vitest --config vitest.config.ts $VERSION",
+    "lint": "eslint tests/**/*.ts",
+    "compile": "tsc -p ./tsconfig.json --noEmit --skipLibCheck",
+    "gen": "graphql-codegen"
+  },
+  "author": "",
+  "license": "ISC",
+  "dependencies": {
```
Comment (krzysztofzuraw) [inline api-tests/package.json:13]: **suggestion**: I would move those `dependencies` to `devDependencies`

File: api-tests/package.json [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    "@typescript-eslint/eslint-plugin": "5.59.6",
+    "@vitest/ui": "0.31.1",
+    "eslint-plugin-chai-friendly": "0.7.2",
+    "eslint-plugin-formatjs": "4.10.1",
+    "eslint-plugin-no-skip-tests": "1.1.0",
+    "eslint-plugin-prettier": "4.2.1",
+    "eslint-plugin-simple-import-sort": "10.0.0",
+    "prettier": "2.8.8",
+    "typescript": "5.0.4",
+    "vite": "4.3.8",
```
Comment (krzysztofzuraw) [inline api-tests/package.json:31]: **suggestion**: I would add `^` to dependencies

File: .github/workflows/apiTests.yml [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+        working-directory: ./api-tests
+      - run: npm run lint
+        working-directory: ./api-tests
+      - run: npm run api-tests
+        working-directory: ./api-tests
+      - run: npm run graphql-codegen
+        working-directory: ./api-tests
+      - run: npm run compile
```
Comment (krzysztofzuraw) [inline .github/workflows/apiTests.yml:19]: **suggestion**: I think type checking should be run before tests

File: api-tests/package.json [hunk 8]
```diff
@@ -0,0 +1,38 @@
+{
+  "name": "api-tests",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "api-tests": "vitest --config vitest.config.ts $VERSION",
+    "lint": "eslint tests/**/*.ts",
+    "compile": "tsc -p ./tsconfig.json --noEmit --skipLibCheck",
```
Comment (krzysztofzuraw) [inline api-tests/package.json]:
```suggestion
    "check-types": "tsc",
```

File: api-tests/package.json [hunk 9]
```diff
@@ -0,0 +1,38 @@
+{
+  "name": "api-tests",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "api-tests": "vitest --config vitest.config.ts $VERSION",
```
Comment (krzysztofzuraw) [inline api-tests/package.json:7]:
```suggestion
    "test": "vitest",
```
What is `$VERSION`? and do you need it?

File: api-tests/package.json [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    "lint": "eslint tests/**/*.ts",
```
Comment (krzysztofzuraw) [inline api-tests/package.json:8]:
```suggestion
    "lint": "eslint 'src/**/*.@(tsx|ts|jsx|js)' --fix",
```

File: api-tests/package.json
Comment (krzysztofzuraw) [inline api-tests/package.json:1]: **suggestion**: I would add `private: true` to this pkg.json to make sure that this is private pkg and it shouldn't be released into npm

File: api-tests/package.json [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    "compile": "tsc -p ./tsconfig.json --noEmit --skipLibCheck",
+    "gen": "graphql-codegen"
+  },
+  "author": "",
+  "license": "ISC",
```
Comment (krzysztofzuraw) [inline api-tests/package.json:12]:
```suggestion
  "license": "BSD",
```

I would keep the same license as this repo

File: api-tests/package.json [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "api-tests": "vitest --config vitest.config.ts $VERSION",
+    "lint": "eslint tests/**/*.ts",
+    "compile": "tsc -p ./tsconfig.json --noEmit --skipLibCheck",
+    "gen": "graphql-codegen"
+  },
+  "author": "",
```
Comment (krzysztofzuraw) [inline api-tests/package.json:11]:
```suggestion
  "author": "Saleor Commerce",
```

File: api-tests/package.json [hunk 13]
```diff
@@ -0,0 +1,38 @@
+{
+  "name": "api-tests",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
```
Comment (krzysztofzuraw) [inline api-tests/package.json:5]:
```suggestion
  "main": "N/A",
```

File: api-tests/package.json [hunk 14]
```diff
@@ -0,0 +1,38 @@
+{
+  "name": "api-tests",
+  "version": "1.0.0",
+  "description": "",
```
Comment (krzysztofzuraw) [inline api-tests/package.json:4]:
```suggestion
  "description": "Saleor API tests",
```

File: api-tests/package.json [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+  "main": "index.js",
+  "scripts": {
+    "api-tests": "vitest --config vitest.config.ts $VERSION",
+    "lint": "eslint tests/**/*.ts",
+    "compile": "tsc -p ./tsconfig.json --noEmit --skipLibCheck",
+    "gen": "graphql-codegen"
```
Comment (krzysztofzuraw) [inline api-tests/package.json]:
```suggestion
    "generate": "graphql-codegen"
```

File: api-tests/codegen.ts [hunk 16]
```diff
@@ -0,0 +1,17 @@
+import type { CodegenConfig } from '@graphql-codegen/cli'
+
+const config: CodegenConfig = {
+  schema: 'https://master.staging.saleor.cloud/graphql/',
```
Comment (krzysztofzuraw) [inline api-tests/codegen.ts:4]: **suggestion**: I think this should be parametrized

File: api-tests/.eslintrc.json [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+  "plugins": [
+    "@typescript-eslint",
+    "import",
+    "formatjs",
+    "no-only-tests",
+    "no-skip-tests",
+    "vitest"
+  ],
+  "rules": {
+    "@typescript-eslint/adjacent-overload-signatures": "error",
```
Comment (krzysztofzuraw) [inline api-tests/.eslintrc.json:24]: **question**: do you need all those rules? maybe you can use those provided by default eslint plugins?

File: api-tests/tsconfig.json [hunk 18]
```diff
@@ -0,0 +1,35 @@
+{
+  "$schema": "https://json.schemastore.org/tsconfig.json",
+  "compilerOptions": {
```
Comment (krzysztofzuraw) [inline api-tests/tsconfig.json:3]: please add `strict: true` here to enable strict checks

https://www.typescriptlang.org/tsconfig#strict

File: api-tests/tsconfig.json [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+    "baseUrl": ".",
+    "strictNullChecks": false,
```
Comment (krzysztofzuraw) [inline api-tests/tsconfig.json:5]:
```suggestion
```

File: api-tests/tsconfig.json [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+    "esModuleInterop": true,
+    "jsx": "react",
```
Comment (krzysztofzuraw) [inline api-tests/tsconfig.json:7]:
```suggestion
```

File: api-tests/tsconfig.json [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+    "lib": [
+      "es2020",
+      "dom",
+      "esnext"
+    ],
+    "skipLibCheck": true,
+    "sourceMap": true,
```
Comment (krzysztofzuraw) [inline api-tests/tsconfig.json:14]:
```suggestion
    "sourceMap": false,
```

File: api-tests/tsconfig.json [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+    "module": "es2020",
+    "moduleResolution": "node",
+    "noUnusedLocals": true,
+    "noUnusedParameters": true,
+    "downlevelIteration": true,
+    "noImplicitThis": true,
+    "alwaysStrict": true,
+    "strictBindCallApply": true,
+    "useUnknownInCatchVariables": true,
+    "paths": {
```
Comment (krzysztofzuraw) [inline api-tests/tsconfig.json:25]:
```suggestion
    "paths": {
      "~/*": ["./src/*"]
    }
```

**Comment by yellowee** [comment]
Closing in favour of 

https://github.com/saleor/saleor/pull/12962

File: api-tests/package.json (hunk 8 above)
Comment (yellowee) [reply api-tests/package.json]: The noEmmit is still required, to not generate js files.

**Pull request closed without merging**